common/task/priorityTaskProcessor.go

```diff
@@ -0,0 +1,155 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// PriorityTaskProcessorOptions configs PriorityTaskProcessor
+	PriorityTaskProcessorOptions struct {
+		WorkerCount int
+		RetryPolicy backoff.RetryPolicy
+	}
+
+	priorityTaskProcessorImpl struct {
+		scheduler PriorityScheduler
+
+		status       int32
+		workerWG     sync.WaitGroup
+		logger       log.Logger
+		metricsScope metrics.Scope
+		options      *PriorityTaskProcessorOptions
+	}
+)
+
+// NewPriorityTaskProcessor creates a new PriorityTaskProcessor
+func NewPriorityTaskProcessor(
+	scheduler PriorityScheduler,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	options *PriorityTaskProcessorOptions,
+) PriorityTaskProcessor {
+	return &priorityTaskProcessorImpl{
+		scheduler:    scheduler,
+		status:       common.DaemonStatusInitialized,
+		logger:       logger,
+		metricsScope: metricsClient.Scope(metrics.PriorityTaskProcessingScope),
+		options:      options,
+	}
+}
+
+func (p *priorityTaskProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.workerWG.Add(p.options.WorkerCount)
+	for i := 0; i < p.options.WorkerCount; i++ {
+		go p.taskWorker()
+	}
+	p.scheduler.Start()
+	p.logger.Info("Priority task processor started.")
+}
+
+func (p *priorityTaskProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	p.scheduler.Stop()
+	if success := common.AwaitWaitGroup(&p.workerWG, time.Minute); !success {
+		p.logger.Warn("Priority task processor timedout on shutdown.")
+	}
+	p.logger.Info("Priority task processor shutdown.")
+}
+
+func (p *priorityTaskProcessorImpl) Submit(task Task, priority int) error {
+	p.metricsScope.IncCounter(metrics.PriorityTaskSubmitRequest)
+	sw := p.metricsScope.StartTimer(metrics.PriorityTaskSubmitLatency)
+	defer sw.Stop()
+
+	return p.scheduler.Schedule(task, priority)
+}
+
+func (p *priorityTaskProcessorImpl) taskWorker() {
+	defer p.workerWG.Done()
+
+	for {
+		task := p.scheduler.Consume()
+		if task == nil {
+			return
+		}
+
+		p.executeTask(task.(Task))
+	}
+}
+
+func (p *priorityTaskProcessorImpl) executeTask(task Task) {
+	sw := p.metricsScope.StartTimer(metrics.PriorityTaskTaskProcessingLatency)
+	defer sw.Stop()
+
+	op := func() error {
+		if err := task.Execute(); err != nil {
+			return task.HandleErr(err)
+		}
+		return nil
+	}
+
+	isRetryable := func(err error) bool {
+		if p.isStopped() {
+			return false
```

Comment:
i would argue the following:

1. a task should not have the ability to occupy a coroutine / thread and execute forever (i.e. retry for ever)
2. a task can have additional function `ScheduleRetry()` which tells the component which load the task to do a re-dispatching
3. there shall be no retry within this component

basically here, the only things to be done:
1. poll a task
2. execute (check shutdown)
3. 
* if no err -> ack
* if err -> check retry err 
** retry -> schedule retry
** no retry -> nack